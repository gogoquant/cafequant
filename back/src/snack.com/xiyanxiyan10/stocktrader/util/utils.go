package util

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
)

// File2Map convert file json to map[string]string
func File2Map(path string) (map[string]string, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	content, err := ioutil.ReadAll(f)
	if err != nil {
		return nil, err
	}
	m := make(map[string]string)
	err = json.Unmarshal([]byte(content), &m)
	if err != nil {
		return nil, err
	}
	return m, nil
}

// SafefloatDivide ...
func SafefloatDivide(lft, rht float64) float64 {
	if rht == 0.0 {
		return lft
	}
	return lft / rht
}

func converError(val interface{}, t string) error {
	return fmt.Errorf("conver error, the %T{%v} can not conver to a %v", val, val, t)
}

// DeepCopyStruct ...
func DeepCopyStruct(source, target interface{}) {
	data, _ := json.Marshal(source)
	json.Unmarshal(data, target)
}

func stringToBool(val string) (bool, error) {
	switch val {
	case "1", "t", "T", "true", "TRUE", "True", "ok", "OK", "yes", "YES":
		return true, nil
	case "0", "f", "F", "false", "FALSE", "False", "":
		return false, nil
	}
	return false, converError(val, "bool")
}

// Struct2Json ...
func Struct2Json(m interface{}) string {
	dataType, _ := json.Marshal(m)
	dataString := string(dataType)
	return dataString
}
