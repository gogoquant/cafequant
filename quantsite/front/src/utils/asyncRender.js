import React from 'react';
import Loading from 'components/Loading';

const asyncRender = getComponent =>
  class AsyncComponent extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        Component: null,
      };
    }

    async componentDidMount() {
      try {
        const { default: Component } = await getComponent();
        // eslint-disable-next-line react/no-did-mount-set-state
        this.setState({
          Component,
        });
      } catch (error) {
        return new Error(error);
      }
    }

    render() {
      const { Component: C } = this.state;
      return C ? <C {...this.props} /> : <Loading />;
    }
  };

export default asyncRender;
