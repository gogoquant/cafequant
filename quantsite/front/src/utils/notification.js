import { notification } from 'antd';

/**
 * notification generator
 * @param {string} type - notification type
 * @returns {function(string, string): void}
 */
const generator = type =>
  /**
   * @param {string} message - notification message
   * @param {string} description - notification description
   */
  (message, description) =>
    notification[type]({
      message,
      description,
    });

const typeMaps = ['success', 'info', 'warning', 'error'];

/**
 * notification
 * Four types: success, info, warning, error
 * @example
 * notification.success(message, description)
 * @type {{object}}
 * @property success - success notification
 * @property info - info notification
 * @property warning - warning notification
 * @property error - error notification
 */
const notificationUtils = {};
typeMaps.forEach(item => {
  notificationUtils[item] = generator(item);
});

export default notificationUtils;
