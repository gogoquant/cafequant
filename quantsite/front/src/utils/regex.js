export const tokenRegex = /^[a-f\d]{8}(-[a-f\d]{4}){3}-[a-f\d]{12}$/;
export const htmlElementRegex = /<[^>]*>/g;
export const editRouteRegex = /^\/topic\/[a-f\d]{24}\/edit$/;
export const userRouteRegex = /^\/user\/\w+$/;
export const loginRouteRegex = /^\/login$/;
