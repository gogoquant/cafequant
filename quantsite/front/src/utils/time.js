import moment from 'moment';

export const fromNow = time => moment(time).fromNow();
