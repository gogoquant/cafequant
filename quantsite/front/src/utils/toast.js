import { message } from 'antd';

/**
 * global toast generator
 * @param {string} type - toast type
 * @returns {function(string): void}
 */
const generator = type =>
  /**
   * @param {string} msg - toast msg
   */
  msg => message[type](msg);

const typeMaps = ['success', 'info', 'warning', 'error'];

/**
 * global toast
 * Four types: success, info, warning, error
 * @example
 * toastUtils.success(msg)
 * @type {{object}}
 * @property success - success toast
 * @property info - info toast
 * @property warning - warning toast
 * @property error - error toast
 */
const toastUtils = {};
typeMaps.forEach(item => {
  toastUtils[item] = generator(item);
});

export default toastUtils;
