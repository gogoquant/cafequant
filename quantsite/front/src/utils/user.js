export const saveUserLocal = userData =>
  localStorage.setItem('user', JSON.stringify(userData));

export const saveUserSession = userData =>
  sessionStorage.setItem('user', JSON.stringify(userData));

export const removeUser = () => {
  localStorage.removeItem('user');
  sessionStorage.removeItem('user');
};

export const getUser = () => {
  const user = localStorage.getItem('user') || sessionStorage.getItem('user');
  return JSON.parse(user);
};
