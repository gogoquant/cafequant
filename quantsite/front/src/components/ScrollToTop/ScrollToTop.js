import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { scrollUtils } from 'utils';

export class ScrollToTopComponent extends React.Component {
  componentDidUpdate(prevProps) {
    if (
      this.props.location !== prevProps.location &&
      this.props.history.action === 'PUSH'
    ) {
      scrollUtils.scrollTo(0);
    }
  }

  render() {
    return this.props.children;
  }
}

ScrollToTopComponent.propTypes = {
  location: PropTypes.object,
  history: PropTypes.shape({
    action: PropTypes.string,
  }),

  children: PropTypes.node.isRequired,
};

export default withRouter(ScrollToTopComponent);
