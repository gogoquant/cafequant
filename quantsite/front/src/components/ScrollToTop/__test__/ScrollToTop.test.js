import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { scrollUtils } from 'utils';
import ScrollToTop, { ScrollToTopComponent } from '../ScrollToTop';

describe('<ScrollToTop />', () => {
  const props = { location: { pathname: '/' }, history: { action: 'PUSH' } };
  const ScrollWrapper = () => (
    <ScrollToTop {...props}>
      <div>Hi</div>
    </ScrollToTop>
  );

  it('should render children', () => {
    const tree = mount(
      <MemoryRouter>
        <ScrollWrapper />
      </MemoryRouter>,
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('should call scrollUtils scrollTo()', () => {
    const spy = sinon.spy(scrollUtils, 'scrollTo');
    const wrapper = shallow(
      <ScrollToTopComponent {...props}>
        <div>Hi</div>
      </ScrollToTopComponent>,
    );
    wrapper.setProps({ location: { pathname: '/user/admin' } });
    expect(spy.calledOnce).toBeTruthy();

    wrapper.setProps({ history: { action: 'PUSH' } });
    expect(spy.calledOnce).toBeTruthy();
  });
});
