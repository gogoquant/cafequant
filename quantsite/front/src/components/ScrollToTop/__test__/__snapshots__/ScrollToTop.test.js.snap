// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollToTop /> should render children 1`] = `
<div>
  Hi
</div>
`;
